Add answers to questions about communicating with external services

brittag opened the pull request:
Customers and prospective customers often ask us how to securely connect back to their existing services. 

Changes:
* Begin the page with a reader-centered plain language overview of what you can do.
* Add information about authentication options.
* Explain why the egress IPs are on this page.

These changes are based on questions from three prospective/current partner agencies:
* See [this email from 8/29/17](https://docs.google.com/document/d/1faf7V56mWoY0RquiaRoWYGTo3lx9bQcTMYYE5NvsiT4/edit#heading=h.qgtqz36ehcjs) for the basis of the authentication options description. (Thanks to @pburkholder and @cnelson for helping with that.)
* See [this conversation from 6/29/17 and 7/5/17](https://docs.google.com/document/d/1tSPw-jR_sKOX8_75ikoSYTqnnN2OXSrzMkF285K8Z58/edit#) for why I put in an explanation of the basic concepts and the public IP addresses.
* See [this conversation from 8/25/17](https://docs.google.com/document/d/1lldSrSWPmI1lfByBeo2E1IWQw7H2vGU_9HC_gcb_awk/edit) for additional questions on this topic.

This needs thorough technical review. I would also welcome adding more details to this page to better answer the questions in those notes docs.

Commit:
Add examples of auth with external services

Commit:
Expand introduction in reader-centered terms

Commit:
Add reassurance about seeing public IPs

cnelson commented on content/docs/apps/static-egress.md:
```diff
@@ -5,15 +5,22 @@ menu:
 
+Authentication options:
+
+* **cloud.gov client application connecting to external services:** If your application on cloud.gov needs to be a client of an existing server application outside of cloud.gov, your application on cloud.gov can have a TLS certificate that the server application validates, or you can use application-level authentication (such as Authorization headers). In addition, you can configure outside servi ...
```
> You can accept HTTPS and TCP connections.

I'm not sure about this line.  At minimum I'd say "You can make" rather than you "You can accept", but it feels weird in general because we don't restrict outbound traffic in any way.  You can send UDP traffic, you can use any layer-7 protocol including but not limited to HTTPS.

Maybe it's better to say "there are no restrictions on outbound connections" or something like that?

cnelson commented on content/docs/apps/static-egress.md:
```diff
@@ -5,15 +5,22 @@ menu:
+Authentication options:
+
+* **cloud.gov client application connecting to external services:** If your application on cloud.gov needs to be a client of an existing server application outside of cloud.gov, your application on cloud.gov can have a TLS certificate that the server application validates, or you can use application-level authentication (such as Authorization headers). In addition, you can configure outside servi ...
+* **cloud.gov server application accepting external client connections:** If you have a server application on cloud.gov that needs to secure connections from a client application outside of cloud.gov, your server application can limit inbound IP ranges using a [user-provided service instance](https://docs.cloudfoundry.org/devguide/services/user-provided.html) of a [route service](https://docs.clo ...
```
> to authenticate with a client application outside of cloud.gov

I think it'd be better to say "to authenticate clients connecting to your server application".  "Authenticate with" makes it sound like the cloud-gov application is the client here when it's actually a server.

pburkholder committed:
Accept Chris's suggestions

brittag commented on content/docs/apps/static-egress.md:
```diff
@@ -15,8 +15,8 @@ You can set up your cloud.gov application to communicate with any external servi
 
-* **cloud.gov client application connecting to external services:** If your application on cloud.gov needs to be a client of an existing server application outside of cloud.gov, your application on cloud.gov can have a TLS certificate that the server application validates, or you can use application-level authentication (such as Authorization headers). In addition, you can configure outside servi ...
-* **cloud.gov server application accepting external client connections:** If you have a server application on cloud.gov that needs to secure connections from a client application outside of cloud.gov, your server application can limit inbound IP ranges using a [user-provided service instance](https://docs.cloudfoundry.org/devguide/services/user-provided.html) of a [route service](https://docs.clo ...
+* **cloud.gov client application connecting to external services:** If your application on cloud.gov needs to be a client of an existing server application outside of cloud.gov, your application on cloud.gov can have a TLS certificate that the server application validates, or you can use application-level authentication (such as Authorization headers). In addition, you can configure outside servi ...
```
How about "There are no restrictions on outbound connections -- for example, you can make HTTPS, TCP, and UDP connections." ? I think we will have readers that are looking for the keyword "TCP", based on [this inquiry email](https://docs.google.com/document/d/1faf7V56mWoY0RquiaRoWYGTo3lx9bQcTMYYE5NvsiT4/edit).

brittag commented on content/docs/apps/static-egress.md:
```diff
@@ -15,8 +15,8 @@ You can set up your cloud.gov application to communicate with any external servi
-* **cloud.gov client application connecting to external services:** If your application on cloud.gov needs to be a client of an existing server application outside of cloud.gov, your application on cloud.gov can have a TLS certificate that the server application validates, or you can use application-level authentication (such as Authorization headers). In addition, you can configure outside servi ...
-* **cloud.gov server application accepting external client connections:** If you have a server application on cloud.gov that needs to secure connections from a client application outside of cloud.gov, your server application can limit inbound IP ranges using a [user-provided service instance](https://docs.cloudfoundry.org/devguide/services/user-provided.html) of a [route service](https://docs.clo ...
+* **cloud.gov client application connecting to external services:** If your application on cloud.gov needs to be a client of an existing server application outside of cloud.gov, your application on cloud.gov can have a TLS certificate that the server application validates, or you can use application-level authentication (such as Authorization headers). In addition, you can configure outside servi ...
+* **cloud.gov server application accepting external client connections:** If you have a server application on cloud.gov that needs to secure connections from a client application outside of cloud.gov, your server application can limit inbound IP ranges using a [user-provided service instance](https://docs.cloudfoundry.org/devguide/services/user-provided.html) of a [route service](https://docs.clo ...
```
In the sentence *"In the current cloud.gov architecture, your server application on cloud.gov cannot use protocol-level authentication, such as a TLS certificate, to authenticate clients connecting to a cloud.gov server application."*, we now have both "your server application on cloud.gov" and "a cloud.gov server application" - we should use the same phrasing for both since they both refer to the same thing.

pburkholder committed:
Rework section on extern client connections

Moved "Your server application on cloud.gov can only serve HTTPS requests, not other TCP or UDP protocols" to opening of that paragraph. Used "your server application on cloud.gov" more consistently.

Commit:
Fixed double parentheses

khandelwal commented on content/docs/apps/static-egress.md:
```diff
@@ -5,15 +5,22 @@ menu:
 ---
 
-If your cloud.gov-hosted application needs access to an external service, that service must be accessible over the internet. All traffic from cloud.gov applications will transit the internet.
+You can set up your cloud.gov application to communicate with any external service that is accessible over the internet. For example, you can integrate your application with services in your agency infrastructure or with APIs from commercial providers.
```
I don't like the phrase 'cloud.gov application'. Applications are hosted on cloud.gov, but are otherwise independent entities.

brittag replied:
I see how that's ambiguous! I revised it for clarity.

khandelwal commented on content/docs/apps/static-egress.md:
```diff
@@ -5,15 +5,22 @@ menu:
 
+Authentication options:
+
+* **cloud.gov client application connecting to external services:** If your application on cloud.gov needs to be a client of an existing server application outside of cloud.gov, your application on cloud.gov can have a TLS certificate that the server application validates, or you can use application-level authentication (such as Authorization headers). In addition, you can configure outside servi ...
```
Could these two paragraphs benefit from a simple diagram in addition to the text? Otherwise, it takes a few readings to grok (at least for me).

brittag replied:
A diagram sounds great to me! I don't know how to model this in a diagram though.

khandelwal commented on content/docs/apps/static-egress.md:
```diff
@@ -5,15 +5,22 @@ menu:
 ## Restricting network exposure of an external service 
 
-Some service providers filter clients based on IP ranges to minimize the surface area of their service on the open internet. If your service provider requires this, you can provide the cloud.gov egress addresses below to the people who handle network security for the service. 
+Some service providers filter clients based on IP ranges to minimize the surface area of their service on the open internet. If your service provider requires this, you can provide the [cloud.gov egress addresses](#cloud-gov-egress-ranges) to the people who handle network security for the service. 
```
'people who handle network security' - is there a name for these people? otherwise this sounds a bit awkward.

brittag replied:
I figure that the formal name would vary a lot depending on the agency - open to suggestions here for something that makes sense.

khandelwal commented on content/docs/apps/static-egress.md:
```diff
@@ -24,4 +31,4 @@ All traffic from cloud.gov-hosted applications originates from these IP ranges:
 * `52.222.123.172/32`
 
-
+We publish these egress (outbound) IP ranges because that information is already public if you look at any cloud.gov application. Filtering by egress IP range is only suitable in *addition* to authentication and authorization between the client and server applications.
```
What does 'if you look at any cloud.gov application' mean?

brittag replied:
Changed it to "inspect". I don't know the correct technical way to say "you can look at the IP address of a website by looking at the IP address of the website"...

Commit:
More specific about application ownership
